### DISCUSSION TITLE
Does this only work inside views?

### DISCUSSION THREAD

**Discussion by gerdemb** [orchetect/SettingsAccess#10] [Q&A] [answered]
First of all, thanks so much for this project! I am writing my first app and failed my very first app review because of this issue...

I have a MenuBarExtra app. When the app is opened for the first time, the user needs to apply some settings so I want to display the settings view. After much trial-an-error I setup an ApplicationDelegate like this:

```
struct MyApp: App {
    @NSApplicationDelegateAdaptor private var applicationDelegate: ApplicationDelegate
```

and then programmatically opened the Settings using `NSApp.sendAction(Selector(("showSettingsWindow:")), to: nil, from: nil)` in the ApplicationDelegate function `applicationDidFinishLaunching()`. Now that doesn't work in MacOS 14.

What I'm trying to understand is if this library `SetttingsAccess` is designed to work inside of the `ApplicationDelegate` which I think is conceptually outside of a `View`. (Please excuse my imprecise terminology and I hope this makes sense). I tried a few things, but couldn't get it to work. My impression is that this will only work inside a View?

I remember experimenting with using `onAppear()` on various views, but could never get it to catch properly when the app started and I ended up using the ApplicationDelegate. Are there some other strategies I could try here?

**Comment by orchetect** [comment]
Yes, SettingsAccess only works within a SwiftUI context because the mechanism it relies on (SettingsLink) is a SwiftUI view itself.

That's 100% a limitation imposed by Apple. There are some creative ways to get around it, but there is no carte blanche way to invoke Settings from a global scope.

That said, I have spent a lot of time trying to figure out the way to just run some arbitrary code at app launch in a SwiftUI app - which seems to be your specific issue. It's surprisingly not straightforward, and it's possible to make a lot of brittle mistakes.

What I settled on was this solution. Using a custom Scene modifier, attaching that modifier to a Scene (such as `Settings`, `WindowGroup`, `Window`, or `MenuBarExtra`) will cause the code closure to execute at app launch when SwiftUI evaluates the App struct's `body`.

```swift
extension Scene {
    public func onSceneBody(_ block: @escaping () -> Void) -> some Scene {
        DispatchQueue.main.async { block() }
        return self
    }
}
```

Then you use it like this. It only needs to be attached to one scene and doesn't really matter which one it is. For example:

```swift
var body: some Scene {
    Settings {
        ContentView()
    }
    .onSceneBody {
        onAppLaunch()
    }
}

private func onAppLaunch() {
    if /* ... */ { openSettings() }
}
```

**Reply by gerdemb** [reply]
Wow, thanks for the super-fast response! I didn't expect that.

I tried your idea of using `onSceneBody()` on the `Scene`, but `openSettings()` throws `settingsLinkNotConnected`. I also noticed that the `onSceneBody()` appears to be called twice. I tried attaching it to my `Settings` scene and my `MenuBarExtra` scene, but the results were the same.

However, this does seem like a good approach to running arbitrary code at APP start-up.

**Reply by orchetect** [reply, marked as answer]
Yeah that sounds right. Since `SettingsLink` is a view, SettingsAccess requires `openSettingsAccess()` to be attached to a view (and can't be a scene), and the view needs to have been created at runtime prior to being able to call `openSettings()`. Then `openSettings()` can be called anywhere within that view hierarchy. With that information, there's a few creative solutions but you're right - the example above may not work as-is.

A possible solution is to attach `openSettingsAccess()` to the label view within your `MenuBarExtra`, and make sure `MenuBarExtra` is the first scene in your `App`'s `body`. It's also not foolproof, as edge cases may cause the view to not be created - such as the user removing the menubar extra from the menubar, or the menubar extra being hidden under a MacBook Pro's camera notch when there are too many menubar extras to fit on screen.

In a nutshell:

```swift
@main struct MyApp: App {
    var body: some Scene {
        MenuBarExtra {
            MenuView() // menu contents
        } label: {
            MenuLabelView() // menubar button label
                .openSettingsAccess()
        }
    
        Settings {
            SettingsView()
        }
    }
}

struct MenuLabelView: View {
    @Environment(\.openSettings) private var openSettings

    var body: some View {
        Text("Menu") // can be image, text, or both
            .onAppear { onAppLaunch() }
    }

    private func onAppLaunch() {
        // must be scheduled so openSettingsAccess() can run before openSettings() is called
        DispatchQueue.main.async {
            if /* ... */ { try? openSettings() }
        }
    }
}
```

**Reply by orchetect** [reply]
Of course the nuclear option is to not use the `Settings` scene at all and just roll your own window using AppKit. But yeah, Apple is really not making things easy to do it the "right way". Please submit a radar to Apple too (and quote radar FB11984872) - the more of us that do, the more they might consider making this easier on us.

**Reply by gerdemb** [reply]
Your work-around of attaching `openSettingsAccess()` to the label of `MenuBarExtra` is working for me. I really appreciate your advice and assistance.

More generally, as a newbie app developer I don't understand how I could have avoided this problem. I don't recall seeing any message that using `showSettingsWindow` was going to be deprecated and having it hard-fail with no backwards compatibility on macOS 14 was a complete surprise. Does Apple have some kind of communication channel with developers or release notes to announce these kind of breaking changes? Are we expected to immediately upgrade and test our apps on the latest macOS version when it's released to catch these kinds of problems?

Anyway, I will submit a radar to Apple flagging this issue and quote your radar as well.

Thanks again for writing this super-helpful library and for your personal assistance on my particular use-case.

**Reply by orchetect** [reply]
This is an unusual case where Apple chose not to show a yellow compiler warning (which would show up regardless what macOS version is used to either compile or run the app). Instead, they show a purple runtime warning - only when running on macOS 14, and only at the time when the `sendAction` method is called.

So yes, a lot of developers will miss this until an end-user discovers Settings no longer opens for them on macOS 14. Or if the dev happens to try it for themselves.

My guess is that Apple couldn't add a compiler check for the specific selectors and had to implement it as a runtime check.

> having it hard-fail with no backwards compatibility on macOS 14 was a complete surprise

Yes, it's not only deprecated, but fully obsoleted which is also unusual. Typically there is a grace period before something becomes fully obsoleted. And the runtime-only warning adds insult to injury.
